config/config.go

```diff
@@ -1545,94 +1547,125 @@ func migrateConfig(configFile, targetDir string) (string, error) {
-			return errors.New("failed to decrypt config after 3 attempts")
-		}
-		key, err := PromptForConfigKey(false)
+// ReadConfig verifies and checks for encryption and loads the config.
+// Prompts for decryption key, if target data is encrypted.
+// Returns the loaded configuration and whether it was encrypted.
+func ReadConfig(configReader io.Reader, keyProvider func() ([]byte, error)) (*Config, bool, error) {
+	reader := bufio.NewReader(configReader)
+
+	pref, err := reader.Peek(len(EncryptConfirmString))
+	if err != nil {
+		return nil, false, err
+	}
+
+	if !ConfirmECS(pref) {
+		// Read unencrypted configuration
+		decoder := json.NewDecoder(reader)
+		c := &Config{}
+		err = decoder.Decode(c)
+		return c, false, err
+	}
+
+	conf, err := readEncryptedConfWithKey(reader, keyProvider)
+	return conf, true, err
+}
+
+// readEncryptedConf reads encrypted configuration and requests key from provider
+func readEncryptedConfWithKey(reader *bufio.Reader, keyProvider func() ([]byte, error)) (*Config, error) {
+	fileData, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return nil, err
+	}
+	for errCounter := 0; errCounter < maxAuthFailures; errCounter++ {
+		key, err := keyProvider()
 		if err != nil {
 			log.Errorf(log.ConfigMgr, "PromptForConfigKey err: %s", err)
-			errCounter++
 			continue
 		}
 
-		var f []byte
-		f = append(f, fileData...)
-		data, err := c.decryptConfigData(f, key)
+		var c *Config
+		c, err = readEncryptedConf(bytes.NewReader(fileData), key)
 		if err != nil {
-			log.Errorf(log.ConfigMgr, "decryptConfigData err: %s", err)
-			errCounter++
+			log.Error(log.ConfigMgr, "Could not decrypt and deserialise data with given key. Invalid password?", err)
 			continue
 		}
+		return c, nil
+	}
+	return nil, errors.New("failed to decrypt config after 3 attempts")
+}
 
-		err = json.Unmarshal(data, c)
-		if err != nil {
-			if errCounter < maxAuthFailures {
-				log.Error(log.ConfigMgr, "Invalid password.")
-			}
-			errCounter++
-			continue
-		}
-		break
+func readEncryptedConf(reader io.Reader, key []byte) (*Config, error) {
+	c := &Config{}
+	data, err := c.decryptConfigData(reader, key)
+	if err != nil {
+		return nil, err
 	}
-	return nil
+
+	err = json.Unmarshal(data, c)
+	return c, err
 }
 
 // SaveConfig saves your configuration to your desired path
 // prompts for encryption key, if necessary
-func (c *Config) SaveConfig(configPath string, dryrun bool) error {
-	if dryrun {
-		return nil
-	}
-
+func (c *Config) SaveConfig(configPath string) error {
 	defaultPath, _, err := GetFilePath(configPath)
 	if err != nil {
 		return err
 	}
+	writer, err := file.Writer(defaultPath)
+	if err != nil {
+		return err
+	}
+	defer writer.Close()
+	return c.Save(writer, func() ([]byte, error) { return PromptForConfigKey(true) })
+}
 
+// Save saves your configuration to the writer
+// with encryption, if configured
+func (c *Config) Save(configWriter io.Writer, keyProvider func() ([]byte, error)) error {
```

Comment:
This seems like it should be unexported for the same reason of `ReadConfig` and `ReadConfig`, causing confusion

Reply:
again exported intentionally, see previous comment

Reply:
Then I think `SaveConfig` needs to become `SaveConfigToFile`

Reply:
Done